File: examples/client_example.js
const dns = require('dns');
const os = require('os');
const crypto = require('crypto');
const readline = require('readline');

// Configuration
const API_URL = "http://localhost:5000/api/client";
const APP_NAME = "My Node Script";
// Replace with the actual secret from your Dashboard
const APP_SECRET = "YOUR_APP_SECRET_HERE"; 

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

function getHWID() {
  // Simple HWID based on system info
  const info = [
    os.platform(),
    os.arch(),
    os.hostname(),
    os.totalmem()
  ].join('-');
  return crypto.createHash('sha256').update(info).digest('hex');
}

async function apiRequest(endpoint, payload) {
  const response = await fetch(`${API_URL}/${endpoint}`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload)
  });
  return response.json();
}

function prompt(question) {
  return new Promise((resolve) => rl.question(question, resolve));
}

async function main() {
  console.log("=== KeyShield Node.js Client ===");

  // 1. Handshake
  console.log("[*] Initializing session with licensing server...");
  let sessionId;
  try {
    const data = await apiRequest('init', { app_name: APP_NAME, secret: APP_SECRET });
    if (!data.success) {
      console.log(`[-] Handshake Failed: ${data.message}`);
      process.exit(1);
    }
    sessionId = data.session_id;
    console.log(`[+] Initialized! Session ID: ${sessionId.slice(0, 8)}...`);
  } catch (error) {
    console.error("[-] Connection failed:", error.message);
    process.exit(1);
  }

  // 2. Authenticate Key
  const licenseKey = await prompt("Enter your License Key: ");
  const hwid = getHWID();

  console.log("[*] Verifying key...");
  try {
    const loginData = await apiRequest('login', {
      session_id: sessionId,
      key: licenseKey.trim(),
      hwid: hwid
    });

    if (!loginData.success) {
      console.log(`[-] Access Denied: ${loginData.message}`);
      process.exit(1);
    }

    console.log("[+] Access Granted!");
    console.log(`    Expiry Date: ${loginData.expiry}`);
  } catch (error) {
    console.error("[-] Verification error:", error.message);
    process.exit(1);
  }

  // 3. Fetch remote variable
  const varName = "welcome_msg";
  console.log(`[*] Fetching remote variable: ${varName}`);
  try {
    const varData = await apiRequest('var', { session_id: sessionId, name: varName });
    if (varData.success) {
      console.log(`[+] Variable value: ${varData.value}`);
    } else {
      console.log(`[-] Failed to fetch variable: ${varData.message}`);
    }
  } catch (error) {
    console.error("[-] Variable fetch error:", error.message);
  }

  // 4. Remote logging
  try {
    await apiRequest('log', { session_id: sessionId, message: "Node client active." });
  } catch (e) {}

  console.log("[+] KeyShield authorization complete! Running main script code...");
  rl.close();
}

main();

File: examples/client_example.py
import requests
import hashlib
import uuid
import sys

# Configuration
API_URL = "http://localhost:5000/api/client"
APP_NAME = "My Python Script"
# Replace with the actual secret from your Dashboard
APP_SECRET = "YOUR_APP_SECRET_HERE" 

def get_hwid():
    # Simple HWID generation based on machine details
    mac = uuid.getnode()
    return hashlib.sha256(str(mac).encode()).hexdigest()

def main():
    print("=== KeyShield Client Authentication ===")
    
    # 1. Handshake (Initialize Session)
    print("[*] Connecting to server...")
    try:
        response = requests.post(f"{API_URL}/init", json={
            "app_name": APP_NAME,
            "secret": APP_SECRET
        })
        res_data = response.json()
        if not res_data.get("success"):
            print(f"[-] Handshake Failed: {res_data.get('message')}")
            sys.exit(1)
        
        session_id = res_data.get("session_id")
        print(f"[+] Handshake successful. Session ID: {session_id[:8]}...")
        
    except Exception as e:
        print(f"[-] Connection Error: {e}")
        sys.exit(1)

    # 2. Authenticate License Key
    license_key = input("Enter your License Key: ").strip()
    hwid = get_hwid()
    
    print("[*] Validating key...")
    try:
        response = requests.post(f"{API_URL}/login", json={
            "session_id": session_id,
            "key": license_key,
            "hwid": hwid
        })
        res_data = response.json()
        if not res_data.get("success"):
            print(f"[-] Validation Failed: {res_data.get('message')}")
            sys.exit(1)
            
        print("[+] Access Granted!")
        print(f"    Expires: {res_data.get('expiry')}")
        
    except Exception as e:
        print(f"[-] Validation Error: {e}")
        sys.exit(1)

    # 3. Retrieve secure remote variable
    # In KeyAuth/KeyShield, you store secrets on the server so users can't crack/extract them.
    var_name = "welcome_msg"
    print(f"[*] Fetching remote variable: {var_name}")
    try:
        response = requests.post(f"{API_URL}/var", json={
            "session_id": session_id,
            "name": var_name
        })
        res_data = response.json()
        if res_data.get("success"):
            print(f"[+] Variable Value: {res_data.get('value')}")
        else:
            print(f"[-] Failed to fetch variable: {res_data.get('message')}")
    except Exception as e:
        print(f"[-] Fetching Variable Error: {e}")

    # 4. Remote Logging
    print("[*] Sending client activity log...")
    try:
        requests.post(f"{API_URL}/log", json={
            "session_id": session_id,
            "message": "User started the main script operations."
        })
    except:
        pass

    print("[+] Hello from your protected script! Running main code...")

if __name__ == "__main__":
    main()
